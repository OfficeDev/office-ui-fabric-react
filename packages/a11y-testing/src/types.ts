export type PropValue = string | number | boolean | any;
export type Props = { [name: string]: PropValue };
export type AccessibilityBehavior = (props: Props) => any;

export interface Rule {
  stringify: () => string;
}

export interface Slot {
  name: string;
  props: Props[];
  expectAttribute: boolean;
  expectedAttribute: string;
  expectedValue: PropValue;
  description: string;
  hidden: boolean;
  afterEvent: string;
  afterEventData: any[];
  checkEvent: boolean;
}

export interface TestFacade {
  slotExists(slotName: string): boolean;
  attributeExists(slotName: string, attributeName: string): boolean;
  attributeHasValue(slotName: string, attributeName: string, value: PropValue): boolean;
  getAttributeValue(slotName: string, attribute: string): PropValue;
  forProps(props: Props): TestFacade;
  afterEvent(slotName: string, eventName: string, args: any[]): void;
}
