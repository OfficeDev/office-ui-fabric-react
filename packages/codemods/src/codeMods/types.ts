import { SourceFile } from 'ts-morph';

// tslint:disable-next-line: interface-name
export interface CodeModResult {
  success?: boolean;
}

// tslint:disable-next-line: interface-name
export interface Codemod<T = SourceFile> {
  /**
   * The version before which the codemod should run.
   * Must be valid semver.
   * Currently has no impact. Maybe remove?
   */
  version?: string;
  /**
   * A string to help identify the codemod.
   */
  name: string;
  /**
   * The actual function that should be run on any given file.
   * TODO, is there a possibility of codemods that would need to execute over mutiple files?
   */
  run: (file: T) => CodeModResult;
  /**
   * If not enabled, then this mod will not be conisdered to run. Only enable it once it's ready
   * to be applied in real world scenarios
   */
  enabled?: boolean;
}
