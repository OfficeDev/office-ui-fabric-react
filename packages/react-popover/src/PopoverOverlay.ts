export * from './components/PopoverOverlay/index';
