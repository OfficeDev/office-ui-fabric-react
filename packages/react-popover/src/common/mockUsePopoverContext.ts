import { PopoverContextValue, popoverDefaultContext, usePopoverContext } from '../popoverContext';

/**
 * A test utility to mock the usePopoverContext hook that relies on context selector
 * Don't forget to call `jest.mock(**\/popupContext.ts)` while using this
 * @param options Popover context values to set for testing
 */
export const mockPopoverContext = (options: Partial<PopoverContextValue> = {}) => {
  const mockContext: PopoverContextValue = {
    ...popoverDefaultContext,
    ...options,
  };
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (usePopoverContext as jest.Mock).mockImplementation((selector: (context: PopoverContextValue) => any) => {
    return selector(mockContext);
  });
};
