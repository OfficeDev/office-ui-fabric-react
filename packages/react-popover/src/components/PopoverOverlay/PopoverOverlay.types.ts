import * as React from 'react';
import { ComponentProps, ComponentState } from '@fluentui/react-utilities';
import { PopoverContextValue } from '../../popoverContext';

/**
 * PopoverOverlay Props
 */
export interface PopoverOverlayProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {}

/**
 * Names of the shorthand properties in PopoverOverlayProps
 */
export type PopoverOverlayShorthandProps = never;

/**
 * Names of PopoverOverlayProps that have a default value in usePopoverOverlay
 */
export type PopoverOverlayDefaultedProps = never;

/**
 * PopoverOverlay State
 */
export interface PopoverOverlayState
  extends ComponentState<PopoverOverlayProps, PopoverOverlayShorthandProps, PopoverOverlayDefaultedProps>,
    Pick<PopoverContextValue, 'open' | 'mountNode' | 'noArrow' | 'size' | 'brand' | 'inverted'> {
  ref: React.Ref<HTMLElement>;
  /**
   * Ref to the arrow element
   */
  arrowRef?: React.Ref<HTMLDivElement>;

  /**
   * CSS class for the arrow element
   */
  arrowClassName?: string;
}
