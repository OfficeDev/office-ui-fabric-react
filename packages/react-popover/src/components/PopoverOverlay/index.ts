export * from './PopoverOverlay';
export * from './PopoverOverlay.types';
export * from './renderPopoverOverlay';
export * from './usePopoverOverlay';
export * from './usePopoverOverlayStyles';
