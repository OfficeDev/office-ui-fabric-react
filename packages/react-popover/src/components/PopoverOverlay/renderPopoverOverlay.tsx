import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import { Portal } from '@fluentui/react-portal';
import { PopoverOverlayState } from './PopoverOverlay.types';

/**
 * Render the final JSX of PopoverOverlay
 */
export const renderPopoverOverlay = (state: PopoverOverlayState) => {
  const { slots, slotProps } = getSlots(state);

  // TODO should hidden Popovers be supported ?
  if (!state.open) {
    return null;
  }

  return (
    <Portal mountNode={state.mountNode}>
      <slots.root {...slotProps.root}>
        {!state.noArrow && <div ref={state.arrowRef} className={state.arrowClassName} />}
        {state.children}
      </slots.root>
    </Portal>
  );
};
