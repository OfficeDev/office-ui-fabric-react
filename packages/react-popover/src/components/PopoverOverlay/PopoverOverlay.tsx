import * as React from 'react';
import { usePopoverOverlay } from './usePopoverOverlay';
import { PopoverOverlayProps } from './PopoverOverlay.types';
import { renderPopoverOverlay } from './renderPopoverOverlay';
import { usePopoverOverlayStyles } from './usePopoverOverlayStyles';

/**
 * PopoverOverlay component renders react children in a positioned box
 */
export const PopoverOverlay = React.forwardRef<HTMLElement, PopoverOverlayProps>((props, ref) => {
  const state = usePopoverOverlay(props, ref);

  usePopoverOverlayStyles(state);
  return renderPopoverOverlay(state);
});

PopoverOverlay.displayName = 'PopoverOverlay';
