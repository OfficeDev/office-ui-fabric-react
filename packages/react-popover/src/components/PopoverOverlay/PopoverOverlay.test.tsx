import * as React from 'react';
import { PopoverOverlay } from './PopoverOverlay';
import { render, fireEvent } from '@testing-library/react';
import { ReactWrapper } from 'enzyme';
import { isConformant } from '../../common/isConformant';
import { Portal } from '@fluentui/react-portal';
import { mockPopoverContext } from '../../common/mockUsePopoverContext';
import { PopoverOverlayProps } from './PopoverOverlay.types';

jest.mock('../../popoverContext');

describe('PopoverOverlay', () => {
  isConformant({
    // as render test will pass a span tag which is also considered one of the skipped helperComponents
    disabledTests: ['as-renders-html'],
    Component: PopoverOverlay,
    displayName: 'PopoverOverlay',
    requiredProps: { open: true } as PopoverOverlayProps,
    helperComponents: [Portal, 'span'],
  });

  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  beforeEach(() => {
    mockPopoverContext({ open: true });
  });

  // PopoverOverlay is rendered by a Portal so won't be available in the rendered container
  const testid = 'component';

  /**
   * Note: see more visual regression tests for PopoverOverlay in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const { getByTestId } = render(<PopoverOverlay data-testid={testid}>Default PopoverOverlay</PopoverOverlay>);
    expect(getByTestId(testid)).toMatchSnapshot();
  });

  it.each([
    ['onMouseEnter', fireEvent.mouseEnter],
    ['onMouseLeave', fireEvent.mouseLeave],
    ['onKeyDown', fireEvent.keyDown],
  ])('should keep the original %s handler', (handler, triggerEvent) => {
    // Arrange
    const spy = jest.fn();
    const { getByTestId } = render(
      <PopoverOverlay data-testid={testid} {...{ [handler]: spy }}>
        Content
      </PopoverOverlay>,
    );

    // Act
    triggerEvent(getByTestId(testid));

    // Assert
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('should not render when open is false', () => {
    // Arrange
    mockPopoverContext({ open: false });
    const { queryByTestId } = render(<PopoverOverlay data-testid={testid}>Content</PopoverOverlay>);

    // Assert
    expect(queryByTestId(testid)).toBeNull();
  });
});
