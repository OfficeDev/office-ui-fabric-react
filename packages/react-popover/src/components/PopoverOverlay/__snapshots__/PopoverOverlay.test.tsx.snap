// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopoverOverlay renders a default state 1`] = `
<div
  class=""
  data-tabster="{\\"deloser\\":{},\\"modalizer\\":{\\"id\\":\\"modal-9\\",\\"isOthersAccessible\\":true}}"
  data-testid="component"
  role="dialog"
>
  <div
    class=""
  />
  Default PopoverOverlay
</div>
`;
